### DISCUSSION TITLE
Signed bins not hit

### DISCUSSION THREAD

**Discussion by suchi2108** [fvutils/pyvsc#197] [Q&A] [answered]
I have a simple alu, operates on signed inputs a and b and op as operation. The coverage model is defined as below:
![image](https://github.com/fvutils/pyvsc/assets/118723121/c7e8d637-1245-4b21-a7e0-0722b1a99b1a)
I have tried many iterations, but the negative numbers (signed inputs) are never hit. what could be possible reason for this ?
![image](https://github.com/fvutils/pyvsc/assets/118723121/1769dce6-abb5-4ca6-9e93-3d9dd0fffc35)
Also, in the snip above, there are two instances of covergroup (I have just one 'my_covergroup'). From where my_covergroup_1 is coming ?

Any lead would be appreciated.

**Comment by alwilson** [comment]
I haven't played much with the coverage, but copying your covergroup snippet I'm able to get negative numbers to hit those bins. Maybe your constraints or rand variable are aren't correct? Or perhaps how you're sampling isn't right? Linking to more source code could help.

I think the 2nd covergroup might be b/c you created my_covergroup again somewhere.

You can also reference the unit tests on coverage and various uses of covergroups:
https://github.com/fvutils/pyvsc/blob/master/ve/unit/test_cov_examples.py

```python3
import vsc

@vsc.randobj
class my_item(object):
    def __init__(self):
        self.a = vsc.rand_int32_t()
        self.b = vsc.rand_int32_t()
        self.op = vsc.rand_bit_t(3)

@vsc.covergroup
class my_covergroup(object):

    def __init__(self, per_instance=True):
        self.with_sample(
            a=vsc.int32_t(),
            b=vsc.int32_t(),
            op=vsc.bit_t(3)
        )
        self.options.per_instance = per_instance
        self.cp1 = vsc.coverpoint(self.a, bins={
            "bin_0" : vsc.bin([-2147483648, -1300000000]),
            "bin_1" : vsc.bin([-1299999999,  2147483647])
            })
        self.cp2 = vsc.coverpoint(self.b, bins={
            "bin_0" : vsc.bin([-2147483648, -1300000000]),
            "bin_1" : vsc.bin([-1299999999,  2147483647])
            })
        self.cp3 = vsc.coverpoint(self.op, bins={
            "bin_0" : vsc.bin([0, 2]),
            "bin_1" : vsc.bin([3, 7])
            })

my_cg = my_covergroup()
it = my_item()

for _ in range(100):
    it.randomize()
    my_cg.sample(it.a, it.b, it.op)

vsc.report_coverage(details=True)
```

Output showing all bins hit:
```bash
$ ./cvg.py
TYPE my_covergroup : 100.00%
    CVP cp1 : 100.00%
    Bins:
        bin_0 : 21
        bin_1 : 79
    CVP cp2 : 100.00%
    Bins:
        bin_0 : 21
        bin_1 : 79
    CVP cp3 : 100.00%
    Bins:
        bin_0 : 37
        bin_1 : 63
    INST my_covergroup : 100.00%
        CVP cp1 : 100.00%
        Bins:
            bin_0 : 21
            bin_1 : 79
        CVP cp2 : 100.00%
        Bins:
            bin_0 : 21
            bin_1 : 79
        CVP cp3 : 100.00%
        Bins:
            bin_0 : 37
            bin_1 : 63
```

**Reply by suchi2108** [reply]
Thank you for the reply. I could fix duplication of covergroup instance. 

For the signed bins, I have checked the sampling values (It does have negative values ). And I am only randomizing it using random python module, although I don't see a problem there too. Really not sure what could be the issue here!

**Reply by suchi2108** [reply, marked as answer]
I fixed the issue. So bascially the DUT was sending 2's complement, which was always out of range for negative numbers and hence no bins were hit.
